### ISSUE TITLE
missing a 'gammas_batch' when computing the gradients of 'grad_f' ?

### ISSUE THREAD

**Issue by weivision** [chi0tzp/WarpedGANSpace#3]
Hi,  I am trying to digest your code and find that you might miss a 'gammas_batch' when computing the gradients of 'grad_f' 

 [grad_f = -2 * (alphas_batch * torch.exp(-gammas_batch * (torch.norm(D, dim=2) ** 2).unsqueeze(dim=2)) * D).sum(dim=1)](https://github.com/chi0tzp/WarpedGANSpace/blob/555717a40e065e42830fcbdf0efc14396e3c7046/lib/support_sets.py#L118)

according to Equation (2) in the paper:
![CodeCogsEqn](https://user-images.githubusercontent.com/34066181/144993484-3ad1be6f-af05-43f7-8b74-8803859b6707.gif),
if my understanding is correct.

**Comment by chi0tzp** [comment]
Hi @weivision, thanks for bringing this to my attention! Indeed, `gammas_batch` should be included in the calculation of `grad_f` similarly to the following:
```
grad_f = -2 * (alphas_batch * gammas_batch * torch.exp(-gammas_batch * (torch.norm(D, dim=2) ** 2).unsqueeze(dim=2)) * D).sum(dim=1)
```
This version of the code is after a large refactoring, before I make it public, so it appears I missed this bit -- I believe that pre-trained models have been trained with the correct formula for `grad_f`, but I might be wrong, so I definitely need to check! I'll train everything from scratch and update the models -- I'll keep this issue open until then. 

Thanks again for pointing this out!

**Comment by weivision** [comment]
Hi @chi0tzp,

Many thanks for your confirmation. Actually, I think your idea is novel and should be technically valid. May I ask one more question that puzzles me when I was reading your paper and code? It seems in code the latent codes are trained with the target class of 238. I am wondering whether such latent codes can be applied to other classes, or I have to train new latent codes for other classes specifically.

**Comment by chi0tzp** [comment]
Hi @weivision, in the case of BigGAN, indeed I trained a model for class 238 -- that was in order to be able to compare with other methods that trained linear direction for that class. In general, you would need to trained a new model for any other classes of BigGAN, but it might be possible that certain paths trained on one class (e.g., 238) could be transferred to other classes as well. This is something that we didn't study in the original paper (but is interesting nevertheless).

**Comment by weivision** [comment]
Hi @chi0tzp, thanks for your quick response :-).
